File: sdk/python/src/detail/model_variant.py
# -------------------------------------------------------------------------
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.
# --------------------------------------------------------------------------
from __future__ import annotations

import logging
from typing import Callable, List, Optional

from ..imodel import IModel
from ..exception import FoundryLocalException

from .core_interop import CoreInterop, InteropRequest
from .model_data_types import ModelInfo
from .core_interop import get_cached_model_ids
from .model_load_manager import ModelLoadManager
from ..openai.audio_client import AudioClient
from ..openai.chat_client import ChatClient
from ..openai.embedding_client import EmbeddingClient

logger = logging.getLogger(__name__)


class ModelVariant(IModel):
    """A specific variant of a model (e.g. a particular device type, version, or quantization).

    Implements ``IModel`` and provides download, cache, load/unload, and
    client-creation operations for a single model variant.
    """

    def __init__(self, model_info: ModelInfo, model_load_manager: ModelLoadManager, core_interop: CoreInterop):
        """Initialize a ModelVariant.

        Args:
            model_info: Catalog metadata for this variant.
            model_load_manager: Manager for loading/unloading models.
            core_interop: Native interop layer for Foundry Local Core.
        """
        self._model_info = model_info
        self._model_load_manager = model_load_manager
        self._core_interop = core_interop

        self._id = model_info.id
        self._alias = model_info.alias

    @property
    def id(self) -> str:
        """Unique model variant ID (e.g. ``name:version``)."""
        return self._id

    @property
    def alias(self) -> str:
        """Model alias shared across variants."""
        return self._alias

    @property
    def info(self) -> ModelInfo:
        """Full catalog metadata for this variant."""
        return self._model_info

    @property
    def context_length(self) -> Optional[int]:
        """Maximum context length (in tokens) supported by this variant, or ``None`` if unknown."""
        return self._model_info.context_length

    @property
    def variants(self) -> List[IModel]:
        """A ModelVariant is a single variant, so variants returns itself."""
        return [self]

    def select_variant(self, variant: IModel) -> None:
        """SelectVariant is not supported on a ModelVariant.

        Call ``Catalog.get_model()`` to get an IModel with all variants available.

        :raises FoundryLocalException: Always.
        """
        raise FoundryLocalException(
            f"select_variant is not supported on a ModelVariant. "
            f'Call Catalog.get_model("{self._alias}") to get an IModel with all variants available.'
        )

    @property
    def input_modalities(self) -> Optional[str]:
        """Comma-separated input modalities (e.g. ``"text,image"``), or ``None`` if unknown."""
        return self._model_info.input_modalities

    @property
    def output_modalities(self) -> Optional[str]:
        """Comma-separated output modalities (e.g. ``"text"``), or ``None`` if unknown."""
        return self._model_info.output_modalities

    @property
    def capabilities(self) -> Optional[str]:
        """Comma-separated capability tags (e.g. ``"chat,completion"``), or ``None`` if unknown."""
        return self._model_info.capabilities

    @property
    def supports_tool_calling(self) -> Optional[bool]:
        """Whether this variant supports tool/function calling, or ``None`` if unknown."""
        return self._model_info.supports_tool_calling

    @property
    def is_cached(self) -> bool:
        """``True`` if this variant is present in the local model cache."""
        cached_model_ids = get_cached_model_ids(self._core_interop)
        return self.id in cached_model_ids

    @property
    def is_loaded(self) -> bool:
        """``True`` if this variant is currently loaded into memory."""
        loaded_model_ids = self._model_load_manager.list_loaded()
        return self.id in loaded_model_ids

    def download(self, progress_callback: Callable[[float], None] = None):
        """Download this variant to the local cache.

        Args:
            progress_callback: Optional callback receiving download progress as a
                percentage (0.0 to 100.0).
        """
        request = InteropRequest(params={"Model": self.id})
        if progress_callback is None:
            response = self._core_interop.execute_command("download_model", request)
        else:
            response = self._core_interop.execute_command_with_callback(
                "download_model", request,
                lambda pct_str: progress_callback(float(pct_str))
            )

        logger.info("Download response: %s", response)
        if response.error is not None:
            raise FoundryLocalException(f"Failed to download model: {response.error}")

    def get_path(self) -> str:
        """Get the local file-system path to this variant if cached.

        Returns:
            Path to the model directory.

        Raises:
            FoundryLocalException: If the model path cannot be retrieved.
        """
        request = InteropRequest(params={"Model": self.id})
        response = self._core_interop.execute_command("get_model_path", request)
        if response.error is not None:
            raise FoundryLocalException(f"Failed to get model path: {response.error}")

        return response.data

    def load(self) -> None:
        """Load this variant into memory for inference."""
        self._model_load_manager.load(self.id)

    def remove_from_cache(self) -> None:
        """Remove this variant from the local model cache."""
        request = InteropRequest(params={"Model": self.id})
        response = self._core_interop.execute_command("remove_cached_model", request)
        if response.error is not None:
            raise FoundryLocalException(f"Failed to remove model from cache: {response.error}")


    def unload(self) -> None:
        """Unload this variant from memory."""
        self._model_load_manager.unload(self.id)

    def get_chat_client(self) -> ChatClient:
        """Create an OpenAI-compatible ``ChatClient`` for this variant."""
        return ChatClient(self.id, self._core_interop)

    def get_audio_client(self) -> AudioClient:
        """Create an OpenAI-compatible ``AudioClient`` for this variant."""
        return AudioClient(self.id, self._core_interop)

    def get_embedding_client(self) -> EmbeddingClient:
        """Create an OpenAI-compatible ``EmbeddingClient`` for this variant."""
        return EmbeddingClient(self.id, self._core_interop)

File: sdk/js/src/detail/modelVariant.ts
import { CoreInterop } from './coreInterop.js';
import { ModelLoadManager } from './modelLoadManager.js';
import { ModelInfo } from '../types.js';
import { ChatClient } from '../openai/chatClient.js';
import { AudioClient } from '../openai/audioClient.js';
import { EmbeddingClient } from '../openai/embeddingClient.js';
import { LiveAudioTranscriptionSession } from '../openai/liveAudioTranscriptionClient.js';
import { ResponsesClient } from '../openai/responsesClient.js';
import { IModel } from '../imodel.js';

/**
 * Represents a specific variant of a model (e.g., a specific quantization or format).
 * Contains the low-level implementation for interacting with the model.
 * @internal
 */
export class ModelVariant implements IModel {
    private _modelInfo: ModelInfo;
    private coreInterop: CoreInterop;
    private modelLoadManager: ModelLoadManager;

    constructor(modelInfo: ModelInfo, coreInterop: CoreInterop, modelLoadManager: ModelLoadManager) {
        this._modelInfo = modelInfo;
        this.coreInterop = coreInterop;
        this.modelLoadManager = modelLoadManager;
    }

    /**
     * Gets the unique identifier of the model variant.
     * @returns The model ID.
     */
    public get id(): string {
        return this._modelInfo.id;
    }

    /**
     * Gets the alias of the model.
     * @returns The model alias.
     */
    public get alias(): string {
        return this._modelInfo.alias;
    }

    /**
     * Gets the detailed information about the model variant.
     * @returns The ModelInfo object.
     */
    public get info(): ModelInfo {
        return this._modelInfo;
    }

    /**
     * A ModelVariant is a single variant, so variants returns itself.
     */
    public get variants(): IModel[] {
        return [this];
    }

    /**
     * SelectVariant is not supported on a ModelVariant.
     * Call Catalog.getModel() to get an IModel with all variants available.
     * @throws Error always.
     */
    public selectVariant(_variant: IModel): void {
        throw new Error(
            `selectVariant is not supported on a ModelVariant. ` +
            `Call Catalog.getModel("${this.alias}") to get an IModel with all variants available.`
        );
    }

    public get contextLength(): number | null {
        return this._modelInfo.contextLength ?? null;
    }

    public get inputModalities(): string | null {
        return this._modelInfo.inputModalities ?? null;
    }

    public get outputModalities(): string | null {
        return this._modelInfo.outputModalities ?? null;
    }

    public get capabilities(): string | null {
        return this._modelInfo.capabilities ?? null;
    }

    public get supportsToolCalling(): boolean | null {
        return this._modelInfo.supportsToolCalling ?? null;
    }

    /**
     * Checks if the model variant is cached locally.
     * @returns True if cached, false otherwise.
     */
    public get isCached(): boolean {
        const cachedModels: string[] = JSON.parse(this.coreInterop.executeCommand("get_cached_models"));
        return cachedModels.includes(this._modelInfo.id);
    }

    /**
     * Checks if the model variant is loaded in memory.
     * @returns True if loaded, false otherwise.
     */
    public async isLoaded(): Promise<boolean> {
        const loadedModels = await this.modelLoadManager.listLoaded();
        return loadedModels.includes(this._modelInfo.id);
    }

    /**
     * Downloads the model variant.
     * @param progressCallback - Optional callback to report download progress (0-100).
     */
    public async download(progressCallback?: (progress: number) => void): Promise<void> {
        const request = { Params: { Model: this._modelInfo.id } };
        if (!progressCallback) {
            this.coreInterop.executeCommand("download_model", request);
        } else {
            await this.coreInterop.executeCommandStreaming("download_model", request, (chunk: string) => {
                const progress = parseFloat(chunk);
                if (!isNaN(progress)) {
                    progressCallback(progress);
                }
            });
        }
    }

    /**
     * Gets the local file path of the model variant.
     * @returns The local file path.
     */
    public get path(): string {
        const request = { Params: { Model: this._modelInfo.id } };
        return this.coreInterop.executeCommand("get_model_path", request);
    }

    /**
     * Loads the model variant into memory.
     * @returns A promise that resolves when the model is loaded.
     */
    public async load(): Promise<void> {
        await this.modelLoadManager.load(this._modelInfo.id);
    }

    /**
     * Removes the model variant from the local cache.
     */
    public removeFromCache(): void {
        this.coreInterop.executeCommand("remove_cached_model", { Params: { Model: this._modelInfo.id } });
    }

    /**
     * Unloads the model variant from memory.
     * @returns A promise that resolves when the model is unloaded.
     */
    public async unload(): Promise<void> {
        await this.modelLoadManager.unload(this._modelInfo.id);
    }

    /**
     * Creates a ChatClient for interacting with the model via chat completions.
     * @returns A ChatClient instance.
     */
    public createChatClient(): ChatClient {
        return new ChatClient(this._modelInfo.id, this.coreInterop);
    }

    /**
     * Creates an AudioClient for interacting with the model via audio operations.
     * @returns An AudioClient instance.
     */
    public createAudioClient(): AudioClient {
        return new AudioClient(this._modelInfo.id, this.coreInterop);
    }

    /**
     * Creates an EmbeddingClient for generating text embeddings with the model.
     * @returns An EmbeddingClient instance.
     */
    public createEmbeddingClient(): EmbeddingClient {
        return new EmbeddingClient(this._modelInfo.id, this.coreInterop);
    }

    /**
     * Creates a LiveAudioTranscriptionSession for real-time audio streaming ASR.
     * @returns A LiveAudioTranscriptionSession instance.
     */
    public createLiveTranscriptionSession(): LiveAudioTranscriptionSession {
        return new LiveAudioTranscriptionSession(this._modelInfo.id, this.coreInterop);
    }

    /**
     * Creates a ResponsesClient for interacting with the model via the Responses API.
     * @param baseUrl - The base URL of the Foundry Local web service.
     * @returns A ResponsesClient instance.
     */
    public createResponsesClient(baseUrl: string): ResponsesClient {
        return new ResponsesClient(baseUrl, this._modelInfo.id);
    }
}
